core/src/main/java/org/infinispan/container/DefaultDataContainer.java

```diff
@@ -356,20 +350,55 @@ public void onEntryRemoved(Entry<K, InternalCacheEntry<K, V>> entry) {
       }
    }
 
-   public static class EntryIterator<K, V> implements Iterator<InternalCacheEntry<K, V>> {
+   public class EntryIterator implements Iterator<InternalCacheEntry<K, V>> {
 
       private final Iterator<InternalCacheEntry<K, V>> it;
 
+      private InternalCacheEntry<K, V> next;
+
       EntryIterator(Iterator<InternalCacheEntry<K, V>> it){this.it=it;}
 
+      private InternalCacheEntry<K, V> getNext() {
+         while (it.hasNext()) {
+            InternalCacheEntry<K, V> entry = it.next();
+            long now;
+            if (entry.canExpire() && entry.isExpired((now = timeService.wallClockTime()))) {
```

Comment:
Good point. I guess you could initialize `now` with 0 and only re-read the wall clock time if the current value happens to be 0?